### ISSUE TITLE
Privilege order violation for the future grants on stage

### ISSUE THREAD

**Issue by toadies** [Titan-Systems/titan#81]
I'm getting the following error when trying to set my grants.

Privilege order violation for the future grants on stage. READ should be granted before/simultaneously with WRITE.

Hopefully this can be reproduced.
```
    bp = Blueprint(name="cyberops-infrastructure", dry_run=True)
    database = Database("SOURCES_RAW")
    schema = Schema("EXAMPLE", database=database, comment="Test", owner="EXAMPLE_ROLE")
    role = Role("EXAMPLE_ROLE")    
    bp.add(database, schema, role)
    
    grants = [
        Grant(priv="USAGE", on_schema="SOURCES_DEV.EXAMPLE", to="EXAMPLE_ROLE"),
        GrantOnAll(priv="SELECT", on_type="TABLE", in_type="SCHEMA", in_name="SOURCES_DEV.EXAMPLE", to="EXAMPLE_ROLE", owner=""),
        FutureGrant(priv="SELECT", on_type="TABLE", in_type="SCHEMA", in_name="SOURCES_DEV.EXAMPLE", to="EXAMPLE_ROLE"),
        GrantOnAll(priv="READ", on_type="STAGE", in_type="SCHEMA", in_name="SOURCES_DEV.EXAMPLE", to="EXAMPLE_ROLE", owner=""),
        FutureGrant(priv="READ", on_type="STAGE", in_type="SCHEMA", in_name="SOURCES_DEV.EXAMPLE", to="EXAMPLE_ROLE"),
        GrantOnAll(priv="WRITE", on_type="STAGE", in_type="SCHEMA", in_name="SOURCES_DEV.EXAMPLE", to="EXAMPLE_ROLE", owner=""),
        FutureGrant(priv="WRITE", on_type="STAGE", in_type="SCHEMA", in_name="SOURCES_DEV.EXAMPLE", to="EXAMPLE_ROLE"),
    ]
    
    bp.add(*grants)
    
    plan = bp.plan(session)
    pprint(plan)
    _ = bp.apply(session, plan)
    pprint(_)
```

Outcome
'USE ROLE SECURITYADMIN',
 **'GRANT WRITE ON FUTURE STAGES IN SCHEMA SOURCES_DEV.EXAMPLE TO ROLE '
 'EXAMPLE_ROLE',**
 'USE ROLE SYSADMIN',
 "ALTER SCHEMA SOURCES_RAW.EXAMPLE SET comment = 'Test'",
 'USE ROLE SYSADMIN',
 'GRANT OWNERSHIP ON SCHEMA SOURCES_RAW.EXAMPLE TO ROLE EXAMPLE_ROLE '
 'COPY CURRENT GRANTS',
 'USE ROLE SECURITYADMIN',
 'GRANT USAGE ON SCHEMA SOURCES_DEV.EXAMPLE TO EXAMPLE_ROLE',
 'USE ROLE SECURITYADMIN',
 'GRANT WRITE ON ALL STAGEs IN SCHEMA SOURCES_DEV.EXAMPLE TO ROLE '
 'EXAMPLE_ROLE',
 'USE ROLE SECURITYADMIN',
 'GRANT READ ON FUTURE STAGES IN SCHEMA SOURCES_DEV.EXAMPLE TO ROLE '
 'EXAMPLE_ROLE',
 'USE ROLE SECURITYADMIN',
 'GRANT SELECT ON FUTURE TABLES IN SCHEMA SOURCES_DEV.EXAMPLE TO ROLE '
 'EXAMPLE_ROLE',
 'USE ROLE SECURITYADMIN',
 'GRANT READ ON ALL STAGEs IN SCHEMA SOURCES_DEV.EXAMPLE TO ROLE '
 'EXAMPLE_ROLE',
 'USE ROLE SECURITYADMIN',
 'GRANT SELECT ON ALL TABLEs IN SCHEMA SOURCES_DEV.EXAMPLE TO ROLE '
 'EXAMPLE_ROLE'

**Comment by teej** [comment]
This behavior is new to me. Here are the Snowflake docs on this, for my own future reference:

> Note
> 
> When granting both the READ and WRITE privileges for an internal stage, the READ privilege must be granted before or at the same time as the WRITE privilege.
> 
> When revoking both the READ and WRITE privileges for an internal stage, the WRITE privilege must be revoked before or at the same time as the READ privilege.
> 
> Operating on a stage also requires the USAGE privilege on the parent database and schema.
> 
> A role must be granted or inherit the OWNERSHIP privilege on the object to create a temporary object that has the same name as the object that already exists in the schema.


Reading the sibling issue in the [terraform provider](https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/1593), they worked around this by having the non-read grants depend on the read grant.

This workaround will work in Titan as well. You can use `Resource.requires` to influence the execution order of commands. Something like this should work:

```
def stage_grants(stage_name: str) -> list[Resource]:
    read_on_all = GrantOnAll(priv="READ", on_type="STAGE", in_type="SCHEMA", in_name="SOURCES_DEV.EXAMPLE", to="EXAMPLE_ROLE", owner=""),
    future_read = FutureGrant(priv="READ", on_type="STAGE", in_type="SCHEMA", in_name="SOURCES_DEV.EXAMPLE", to="EXAMPLE_ROLE"),
    write_on_all = GrantOnAll(priv="WRITE", on_type="STAGE", in_type="SCHEMA", in_name="SOURCES_DEV.EXAMPLE", to="EXAMPLE_ROLE", owner=""),
    future_write = FutureGrant(priv="WRITE", on_type="STAGE", in_type="SCHEMA", in_name="SOURCES_DEV.EXAMPLE", to="EXAMPLE_ROLE"),

    write_on_all.requires(read_on_all)
    future_write.requires(read_on_all)
    return [read_on_all, future_read, write_on_all, future_write]
```

I'll repro this locally and have `Blueprint` automatically create the dependency.

**Comment by toadies** [comment]
I can work with that.  Thank you for the feedback.

is this considered a bug or should we close it?

**Comment by teej** [comment]
I consider this a bug, I'll fix it in an upcoming release.

**Comment by toadies** [comment]
Is it possible to include requires within a yaml config file?

**Comment by teej** [comment]
Fixed in [v0.11.0](https://github.com/Titan-Systems/titan/releases/tag/v0.11.0).
